bpo-46649: Propagate Python thread name to PyThreadState

P403n1x87 opened the pull request:
This change allows Python thread objects to propagate their name to the underlying PyThreadState structure to make it easier for tools that rely on the ABI, like sampling profilers, to retrieve this information and provide richer profiling data.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- issue-number: [bpo-46649](https://bugs.python.org/issue46649) -->
https://bugs.python.org/issue46649
<!-- /issue-number -->

JelleZijlstra commented on Modules/_threadmodule.c:
```diff
@@ -1551,6 +1551,40 @@ PyDoc_STRVAR(excepthook_doc,
 
+
+static PyObject*
+thread_set_name(PyObject *self, PyObject *args) {
```
```
 % ./python.exe
Python 3.11.0a5+ (heads/bpo-412569/propagate-thread-name:9f9acf49be, Feb  5 2022, 17:15:00) [Clang 13.0.0 (clang-1300.0.29.3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import _thread
>>> _thread.set_name(123434, "x")
zsh: segmentation fault  ./python.exe
```

This function basically just writes to arbitrary memory, which is a security risk. I think the right fix would be to pass the thread state around as some sort of opaque object instead of a raw int.

P403n1x87 replied:
Good point. `set_name` is not meant to be called by anything else other than `Thread`, but it must be made safe nonetheless! Is there an established pattern for this sort of issue? I'm thinking of making an object, like `_PyThreadState_Reference` to pass from C to Python.

JelleZijlstra commented on Lib/test/test_threading.py:
```diff
@@ -961,6 +962,37 @@ def import_threading():
+        thread.join()
+
+        # Check we can still set the name after the thread terminated
+        thread.name = "foo"
```
This could use some more tests checking that the name you set is actually reflected.

P403n1x87 replied:
This was meant to catch any potential segfaults, since once the thread terminates there isn't a thread state anymore.

JelleZijlstra commented on Include/pystate.h:
```diff
@@ -14,13 +14,20 @@ removed (with effort). */
 
+PyAPI_DATA(PyTypeObject) PyTStateRef_Type;
+
+#define PyTStateRef_Check(op) Py_IS_TYPE(op, &PyTStateRef_Type)
```
I don't think any of this needs to go in a public header file.

P403n1x87 replied:
I've put this here so that `PyTStateRef_Check(op)` could be used in `_testcapi`. But given how simple the definition is I could get rid of it.

P403n1x87 replied:
@JelleZijlstra I can't get this to work if I remove the `PyAPI_DATA` export. Any idea where I could put that to avoid affecting the stable API? I think one thing to consider, though, is that the new `PyTStateRef` objects are surfaced to Python.

JelleZijlstra replied:
I'm not too experienced with the C API, but I am pretty confident we don't want to expose this class in the C API without a strong reason. Generally we're trying to restrict the C API so that it is easier to evolve CPython. For example, changes to the PyThreadState have caused trouble several times. https://devguide.python.org/c-api/ says internal symbols should ideally be in Include/internal.

https://docs.python.org/3/c-api/init.html#c.PyThreadState also says that only the "interp" member of this struct is public, so putting it here isn't enough to allow well-behaved external users of the API to access the thread name. I think you should prefix the new struct members with underscores and add a new, documented and public function `PyThreadState_GetName()` for extensions to access the thread name.

JelleZijlstra commented on Modules/_testcapimodule.c:
```diff
@@ -5770,6 +5770,13 @@ test_tstate_capi(PyObject *self, PyObject *Py_UNUSED(args))
+static PyObject *
+get_thread_name(PyObject *self, PyTStateRef *tsr)
+{
+    assert(PyTStateRef_Check(tsr));
```
Should raise an exception here instead of just asserting (similar to type_get_version above).

JelleZijlstra commented on Modules/_threadmodule.c:
```diff
@@ -1551,6 +1551,102 @@ PyDoc_STRVAR(excepthook_doc,
+    "TStateRef",
+    sizeof(PyTStateRef),
+    0,
+    (destructor)PyObject_Free,          /* tp_dealloc */
```
I believe you can use designated initializers (`.tp_dealloc = `) to avoid all those 0s.

JelleZijlstra reviewed (changes requested):
What happens if you hold on to a `PyTStateRef` object after the corresponding thread has ended? I wasn't able to get a segfault this way, but it looks like I'd be writing into freed memory.

P403n1x87 committed:
bpo-46649: Propagate Python thread name to PyThreadState

This change allows Python thread objects to propagate their name
to the underlying PyThreadState structure to make it easier for
tools that rely on the ABI, like sampling profilers, to retrieve
this information and provide reacher profiling data.

JelleZijlstra commented on Python/pystate.c:
```diff
@@ -783,7 +783,7 @@ init_threadstate(PyThreadState *tstate,
     tstate->datastack_top = NULL;
     tstate->datastack_limit = NULL;
-
+    
```
unnecessary change

JelleZijlstra commented on Python/pystate.c:
```diff
@@ -1034,6 +1034,11 @@ PyThreadState_Clear(PyThreadState *tstate)
 
     Py_CLEAR(tstate->context);
+    
+    if (tstate->ref != NULL)
```
I'm not sure how PyThreadState interacts with the garbage collector. How will the GC know that the thread state ref object is alive?

P403n1x87 replied:
The ref object has a life on its own. The PyThreadState creates the ref object. Once the thread state is deallocated, it sets the actual ref inside the ref object to NULL and decrements its refcount. The ref object then lives on until there are references to it. Otherwise, it is GC'd at the next collection cycle.

JelleZijlstra reviewed (changes requested):
Thanks! Here's a few more comments.

pablogsal reviewed (changes requested):
Unfortunately, I am not convinced this is a good idea. We are moving more and more to a private thread state, and we want to be able to modify its ABI, fields and ordering even between patch releases so I am not comfortable assuring that even more fields can be retrieved via a constant offset from the thread state. Also, adding Python objects to the thread state is possible, but makes destruction more tricky (for example, destroying Python objects must be done with the GIL held, but when the thread state is being teared down from an extension module the GIL may be unlocked).

Additionally, I think this PR is adding quite a lot of extra complexity for very restricted gains.

I am not strongly against this, but unless there is strong consensus that this is clearly a win, I am not confident with this.
